src/test/BasicTests.hs

```diff
@@ -0,0 +1,139 @@
+  it "should transform pattern matching into case expressions" $ do
+    mod1 <-
+      parseTestModule
+      $  "module A where\nlengthL :: [a] -> Int\n"
+      ++ "lengthL [] = 0\nlengthL(_:xs) = 1 + lengthL xs"
+    expected <-
+      parseTestModule
+      $  "module A where\nlengthL :: [a] -> Int\n"
+      ++ "lengthL a0 = case a0 of\n"
+      ++ "  []    -> 0\n"
+      ++ "  a1:a2 -> 1 + lengthL a2"
+    let mod2 = evalPM (processModule mod1) defaultState
+    expected `prettyShouldBe` mod2
+  it "should transform pattern matching in a partial function" $ do
+    mod1 <-
+      parseTestModule
+      $  "module A where\n"
+      ++ "head :: [a] -> a\n"
+      ++ "head (x:xs) = x"
+    let mod2 = evalPM (processModule mod1) defaultState
+    expected <-
+      parseTestModule
+      $  "module A where\n"
+      ++ "head :: [a] -> a\n"
+      ++ "head a0 = case a0 of\n"
+      ++ "  a1 : a2 -> a1\n"
+      ++ "  [] -> undefined"
+    mod2 `prettyShouldBe` expected
+  it "should accept a simple guarded expression" $ do
+    mod1 <-
+      parseTestModule
+      $  "module A where\n"
+      ++ "id :: a -> a\n"
+      ++ "id x | otherwise = x"
+    let mod2 = evalPM (processModule mod1) defaultState
+    expected <-
+      parseTestModule
+      $  "module A where\n"
+      ++ "id :: a -> a\n"
+      ++ "id a0 = let a2 = undefined\n"
+      ++ "            a1 = case a0 of\n"
+      ++ "              a3 -> if otherwise then a3"
+      ++ "                    else a2\n"
+      ++ "         in a1"
+    mod2 `prettyShouldBe` expected
+  it "should accept a more complex guarded function" $ do
+    mod1 <-
+      parseTestModule
+      $  "module A where\n"
+      ++ "useless :: (a -> Bool) -> a -> a -> a\n"
+      ++ "useless p x y | p x       = x\n"
+      ++ "              | otherwise = y"
+    let mod2 = evalPM (processModule mod1) defaultState
+    expected <-
+      parseTestModule
+      $  "module A where\n"
+      ++ "useless :: (a -> Bool) -> a -> a -> a\n"
+      ++ "useless a0 a1 a2 = let\n"
+      ++ "                     a4 = undefined\n"
+      ++ "                     a3 = case a0 of\n"
+      ++ "                       a5 -> case a1 of\n"
+      ++ "                        a6 -> case a2 of\n"
+      ++ "                         a7 -> if a5 a6 then a6\n"
+      ++ "                               else if otherwise then a7\n"
+      ++ "                                    else a4\n"
+      ++ "                   in a3"
+    mod2 `prettyShouldBe` expected
+
+-------------------------------------------------------------------------------
+-- Pretty printing comparison                                                --
+-------------------------------------------------------------------------------
+
+-- | Pretty prints both values and tests whether the resulting strings are
+--   equal modulo whitespace.
+prettyShouldBe :: (Pretty a, Pretty b) => a -> b -> Expectation
+prettyShouldBe x y =
+  let discardWhitespace = unwords . words
```

Comment:
Wouldn't it be useful to disregard line breaks as well?

Reply:
Line breaks are ignored by `discardWhitespace` already.

```haskell
λ> unwords (words "first line\nsecond line")
"first line second line"
```

But that may actually be problematic. For example, the following two programs are obviously not equivalent.

```haskell
data Bar
foo = Foo
```

```haskell
data Bar foo = Foo
```

But `prettyShouldBe` would treat them as equal.

```haskell
λ> unwords (words "data Bar\nfoo = Foo")
"data Bar foo = Foo"
```

In the Free Compiler this is not a problem since IR and Coq both don't really care about whitespace. Solving this problem for Haskell is not trivial. If we extended `discardWhitespace` to take newlines into account, other problems arise. In the following example, it wouldn't be clear which `case` expression the  second alternative belongs to.

```haskell
foo xs ys = case xs of
  [] -> case ys of
    x : xs -> 
```

And even if indentation is preserved, this approach wouldn't work since the following programs wouldn't be equal now.

```haskell
foo x = case x of Foo -> 1
```

```haskell
foo x = case x of
  Foo -> 1
```

I think the only way to solve this issue is by implementing a custom equality for our syntax. In the Free Compiler we do something similar for the IR (see [`FreeC.IR.Similar`](https://freeproving.github.io/free-compiler/docs/master/FreeC-IR-Similar.html)). The advantage of the `similar`ity predicate is that it is compatible with renaming of local variables which we would profit from as well (especially due to fresh variables).

Reply:
Thanks for this detailed explanation. But as we currently parse and pretty print both sides of `shouldBe` anyway, I think we could (at least currently) even remove `discardWhitespace` entirely, because `prettyPrint` should add the same amount of them.
But a custom equality would definitely be useful for more complex tests.

Reply:
If I understand the thumbs up by @just95 correctly, the outcome of this conversation was to remove `discardWhitespace` for the moment.